File: snow crab/size-based model TMB/model_basis.cpp
#include <TMB.hpp>
template<class Type> Type objective_function<Type>::operator()(){
   // Data declarations:
   DATA_VECTOR(x);                        // Size values (n_size).
   DATA_MATRIX(f_imm);                    // Observed frequencies of immature crab (n_year x n_size).
   DATA_MATRIX(f_mat);                    // Observed frequencies of mature crab (n_year x n_size).
   DATA_MATRIX(basis);                    // Instar basis functions.
      
   // Model parameters:
   PARAMETER(intercept_growth);           // Growth intercept parameter.
   PARAMETER(xp_growth);                  // Growth pivot point parameter.
   PARAMETER(log_window_growth);          // Growth transition window parameter.
   PARAMETER_VECTOR(slope_growth);        // Growth slope parameters.
   PARAMETER(log_sigma_growth);           // Growth error parameter.
   
   // Selectivity parameters:                       
   PARAMETER(log_shape_selectivity);      // Shape parameter for cumulative gamma distribution function.
   PARAMETER(log_scale_selectivity);      // Scale parameter for cumulative gamma distribution function.
   
   // Maturation parameters:
   PARAMETER_VECTOR(xp_maturity);         // Maturation pivot point parameters.
   PARAMETER_VECTOR(log_window_maturity); // Maturation transition window parameters.
   PARAMETER(logit_p_mix_maturity);       // Mixing proportion between early and late maturation (logit-scale).
   
   // Skip-moulting parameters:
   PARAMETER(xp_skp);                     // Skip-moulting pivot point parameter.
   PARAMETER(log_window_skp);             // Skip-moulting transition window parameter.
   PARAMETER(logit_p_max_skp);            // Maximum skip-moulting probability (logit-scale).
   
   // Fishing mortality parameters:
   PARAMETER(xp_F);                       // Exploitation rate switching point.
   PARAMETER(log_window_F);               // Exploitation switching point window.
   PARAMETER(log_mu_F);                   // Global exploitation rate.
   PARAMETER_VECTOR(log_dev_F);           // Annual deviations exploitation rate.
   
   // Natural mortality parameters:
   PARAMETER(log_mu_M_imm);               // Global natural mortality rate for immature crab.
   PARAMETER_VECTOR(log_dev_M_imm);       // Annual deviations natural mortality rate for immature crab.
   PARAMETER(log_mu_M_mat);               // Global natural mortality rate for mature crab.
   PARAMETER_VECTOR(log_dev_M_mat);       // Annual deviations natural mortality rate for mature crab.
   
   // Recruitment parameters:                       
   PARAMETER(size_recruitment_mu);           // Mean recruitment size.
   PARAMETER(log_sigma_size_recruitment);    // Error for recruitment annual deviations.
   PARAMETER_VECTOR(size_recruitment);       // Annual deviations from recruitment instar mean size.
   PARAMETER(log_sigma_recruitment_instar);  // Error for recruitment instar.
   PARAMETER(log_scale_recruitment_mu);      // Recruitment scale mean.
   PARAMETER(log_sigma_scale_recruitment);   // Recruitment scale error.
   PARAMETER_VECTOR(log_scale_recruitment);  // Recruitment scale deviations.
   
   // Initial year survey numbers:
   PARAMETER_VECTOR(log_N_imm_reg_beta);  // Basis coefficients for immature crab for year 0.
   PARAMETER_VECTOR(log_N_imm_skp_beta);  // Basis coefficients for immature skip-moulting crab for year 0.
   PARAMETER_VECTOR(log_N_mat_new_beta);  // Basis coefficients for new mature crab for year 0.
   PARAMETER_VECTOR(log_N_mat_old_beta);  // Basis coefficients for old mature crab for year 0. 

   // Error parameters for random effects:
   PARAMETER(log_sigma_dev_F);            // Error for annual exploitation rate immature crab random effect.
   PARAMETER(log_sigma_dev_M_imm);        // Error for annual mortality rate immature crab random effect.
   PARAMETER(log_sigma_dev_M_mat);        // Error for annual mortality rate mature crab random effect.
   
   // Error distribution parameters:
   PARAMETER(log_r_imm);                  // Negative binomial dispersion coefficient for immature crab.
   PARAMETER(log_r_mat);                  // Negative binomial dispersion coefficient for mature crab.  
   
   // Dimensional counters: 
   int n_size = x.size();                 // Number of size categories.
   int n_year = f_imm.rows();             // Number of years of data.
   int n_basis = basis.rows();            // Number of instar basis functions.
   
   // Accumulators and constants:
   Type nll = 0;                          // Initialize negative log-likelihood.
   Type dx = 1.0;                         // Size variable precision.
   
   // Parameter transforms:
   Type p_mix_maturity = 1.0 / (1.0 + exp(-logit_p_mix_maturity));  // Mixing proportion between early and late maturation.
   Type p_max_skp      = 1.0 / (1.0 + exp(-logit_p_max_skp));       // Maximum skip-moulting probability.
   Type r_imm          = exp(log_r_imm); // Negative binomial dispersion coefficient for immature crab.
   Type r_mat          = exp(log_r_mat); // Negative binomial dispersion coefficient for mature crab.
   
   // Random effects:
   nll -= sum(dnorm(size_recruitment, 0.0, exp(log_sigma_size_recruitment), true));
   nll -= sum(dnorm(log_scale_recruitment, 0.0, exp(log_sigma_scale_recruitment), true));
   nll -= sum(dnorm(log_dev_F, 0.0, exp(log_sigma_dev_F), true));
   nll -= sum(dnorm(log_dev_M_imm, 0.0, exp(log_sigma_dev_M_imm), true));
   nll -= sum(dnorm(log_dev_M_mat, 0.0, exp(log_sigma_dev_M_mat), true));
   
   // Smoothed piece-wise linear growth model:
   vector<Type> mu_growth(n_size);
   vector<Type> sigma_growth(n_size);
   vector<Type> phi_growth(n_size);
   vector<Type> k_growth(n_size);
   for (int i = 0; i < n_size; i++){
      // Growth increment:
      mu_growth[i] = intercept_growth +
         slope_growth[0] * x[i] +
         exp(log_window_growth) * (slope_growth[1] - slope_growth[0]) *
         log(1 + exp((x[i] - xp_growth) / exp(log_window_growth)));
      
      // Growth error:
      sigma_growth[i] = exp(log_sigma_growth) * mu_growth[i];
      
      // Gamma parameters:
      phi_growth[i] = (sigma_growth[i] * sigma_growth[i]) / mu_growth[i];  // Gamma scale parameters.
      k_growth[i]   = mu_growth[i] / phi_growth[i];                        // Gamma shape parameters.
   }
   
   // Growth increment matrix:
   matrix<Type> G(n_size, n_size);
   G.fill(0);
   for (int i = 0; i < n_size; i++){
      for (int j = i; j < (n_size-1); j++){
         G(j,i) = pgamma(x[j] - x[i] + (dx / 2.0), k_growth[i], phi_growth[i]) - 
            pgamma(x[j] - x[i] - (dx / 2.0), k_growth[i], phi_growth[i]);
      }
      G(n_size-1,i) = 1.0 - pgamma(x[n_size-1] - x[i] - (dx / 2.0), k_growth[i], phi_growth[i]);
   }
   
   // Evaluate selectivity probabilities:
   vector<Type> p_selectivity = pgamma(x, exp(log_shape_selectivity), exp(log_scale_selectivity));   
   
   // Evaluate maturation probabilities:
   vector<Type> p_maturity_early = 1.0 / (1.0 + exp(-exp(log_window_maturity[0]) * (x - xp_maturity[0])));
   vector<Type> p_maturity_late  = 1.0 / (1.0 + exp(-exp(log_window_maturity[1]) * (x - xp_maturity[1])));
   vector<Type> p_maturity       = (1.0 - p_mix_maturity) * p_maturity_early + p_mix_maturity * p_maturity_late;
   
   // Evaluate skip-moulting probabilties:
   vector<Type> p_skp = p_max_skp / (1 + exp(-exp(log_window_skp) * (x - xp_skp)));
   
   // Evaluate fishery mortality by size and year:
   vector<Type> F_year = 1.0 / (1.0 + exp(-log_mu_F - log_dev_F));
   matrix<Type> F(n_year, n_size);
   for (int i = 0; i < n_year; i++){
      for (int j = 0; j < n_size; j++){
         F(i,j) = F_year[i] / (1+exp(-exp(log_window_F) * (x[j] - xp_F)));
      }
   }
   
   // Evaluate fishery mortality by size and year:
   vector<Type> M_year_imm = 1.0 / (1.0 + exp(-log_mu_M_imm - log_dev_M_imm));
   matrix<Type> M_imm(n_year, n_size);
   vector<Type> M_year_mat = 1.0 / (1.0 + exp(-log_mu_M_mat - log_dev_M_mat));
   matrix<Type> M_mat(n_year, n_size);
   for (int i = 0; i < n_year; i++){
      for (int j = 0; j < n_size; j++){
         M_imm(i,j) = M_year_imm[i];
         M_mat(i,j) = M_year_mat[i];
      }
   }
   
   // Discretize population recruitment: 
   matrix<Type> p_recruitment(n_year,n_size);
   for (int i = 0; i < n_year; i++){
      p_recruitment(i,0) = pnorm(x[0] + 0.5, size_recruitment_mu + size_recruitment[i], exp(log_sigma_recruitment_instar));
      for (int j = 1; j < (n_size-1); j++){
         p_recruitment(i,j) = pnorm(x[j] + 0.5, size_recruitment_mu + size_recruitment[i], exp(log_sigma_recruitment_instar)) - 
                              pnorm(x[j] - 0.5, size_recruitment_mu + size_recruitment[i], exp(log_sigma_recruitment_instar));
      }
      p_recruitment(i, n_size-1) = 1.0 - pnorm(x[n_size-1] - 0.5, size_recruitment_mu + size_recruitment[i], exp(log_sigma_recruitment_instar));
   }
   
   // Scale recruitment: 
   matrix<Type> N_recruitment(n_year, n_size);
   for (int i = 0; i < n_year; i++){
      for (int j = 0; j < n_size; j++){
         N_recruitment(i,j) = exp(log_scale_recruitment_mu + log_scale_recruitment[i]) * p_recruitment(i,j);    
      }
   }
   
   // Initialize population numbers: 
   matrix<Type> N_imm_reg(n_year+1,n_size);
   matrix<Type> N_imm_skp(n_year+1,n_size);
   matrix<Type> N_mat_new(n_year+1,n_size);
   matrix<Type> N_mat_old(n_year+1,n_size);
   N_imm_reg.fill(0);
   N_imm_skp.fill(0);
   N_mat_new.fill(0);
   N_mat_old.fill(0);
   for (int j = 0; j < n_size; j++){
      for (int i = 0; i < n_basis; i++){
         N_imm_reg(0,j) += exp(log_N_imm_reg_beta[i]) * basis(i,j); // Immature crab for year 0.
      }
      // Drop out immature instars:
      for (int i = 4; i < n_basis; i++){
         N_imm_skp(0,j) += exp(log_N_imm_skp_beta[i-4]) * basis(i,j); // Immature skip-moulting crab for year 0.
         N_mat_new(0,j) += exp(log_N_mat_new_beta[i-4]) * basis(i,j); // New mature crab for year 0.
         N_mat_old(0,j) += exp(log_N_mat_old_beta[i-4]) * basis(i,j); // Old mature crab for year 0. 
      }
   }  
   
   // Population dynamics model:
   for (int i = 1; i < (n_year+1); i++){
      for (int j = 0; j < n_size; j++){
         // Skip-moulters:
         N_imm_skp(i,j) = (1.0 - M_imm(i-1,j)) * p_skp[j] * (1.0-p_maturity[j]) * N_imm_reg(i-1,j);  
         
         // Apply growth transition matrix:
         for (int k = 0; k < n_size; k++){
            // Immature new-shelled:
            N_imm_reg(i,k) += (1.0 - M_imm(i-1,j)) * (1.0-p_skp[j]) * (1.0-p_maturity[j]) * N_imm_reg(i-1,j) * G(k,j); 
            
            // Mature new-shelled:
            N_mat_new(i,k) += (1.0 - M_mat(i-1,j)) * ((p_maturity[j] * N_imm_reg(i-1,j)) + N_imm_skp(i-1,j)) * G(k,j);  
         }
         
         // Add recruitment:
         N_imm_reg(i,j) += N_recruitment(i-1,j);
         
         // Mature old-shelled:
         N_mat_old(i,j) = (1.0 - F(i-1,j)) * (1.0 - M_mat(i-1,j)) * (N_mat_new(i-1,j) + N_mat_old(i-1,j)); 
      }
   }
   
   // Survey numbers:
   matrix<Type> n_imm_reg(n_year+1,n_size);
   matrix<Type> n_imm_skp(n_year+1,n_size);
   matrix<Type> n_mat_new(n_year+1,n_size);
   matrix<Type> n_mat_old(n_year+1,n_size);
   matrix<Type> n_imm(n_year+1,n_size);
   matrix<Type> n_mat(n_year+1,n_size);
   for (int i = 0; i < (n_year+1); i++){
      for (int j = 0; j < n_size; j++){
         n_imm_reg(i,j) = p_selectivity[j] * N_imm_reg(i,j);
         n_imm_skp(i,j) = p_selectivity[j] * N_imm_skp(i,j);
         n_mat_new(i,j) = p_selectivity[j] * N_mat_new(i,j);
         n_mat_old(i,j) = p_selectivity[j] * N_mat_old(i,j);
         n_imm(i,j) = n_imm_reg(i,j) + n_imm_skp(i,j);
         n_mat(i,j) = n_mat_new(i,j) + n_mat_old(i,j);
      }
   }
   
   // Likelihood function:
   for (int i = 0; i < n_year; i++){
      for (int j = 0; j < n_size; j++){
         // Immature likelihood:
         nll -= lgamma(f_imm(i,j) + r_imm) - lgamma(r_imm) - lgamma(f_imm(i,j) + 1) + r_imm * log(r_imm) + 
            f_imm(i,j) * log(n_imm(i+1,j)) - (r_imm + f_imm(i,j)) * log(r_imm + n_imm(i+1,j));
         
         // Mature likelihood:
         nll -= lgamma(f_mat(i,j) + r_mat) - lgamma(r_mat) - lgamma(f_mat(i,j) + 1) + r_mat * log(r_mat) + 
            f_mat(i,j) * log(n_mat(i+1,j)) - (r_mat + f_mat(i,j)) * log(r_mat + n_mat(i+1,j));
      }
   }
   
   // ================================= Output =================================
   
   // Growth:
   REPORT(mu_growth);
   REPORT(sigma_growth);
   REPORT(phi_growth);
   REPORT(k_growth);
   REPORT(G);
   
   // Selectivity:
   REPORT(p_selectivity);
   
   // Maturity:
   REPORT(p_maturity);
   REPORT(p_skp);
   
   // Mortality:
   REPORT(F);
   REPORT(F_year);
   REPORT(M_imm);
   REPORT(M_year_imm);
   REPORT(M_mat);
   REPORT(M_year_mat);
   
   // Recruitment:
   REPORT(p_recruitment);
   REPORT(N_recruitment);
   
   // Population numbers:
   REPORT(N_imm_reg);
   REPORT(N_imm_skp);
   REPORT(N_mat_new);
   REPORT(N_mat_old);

   // Survey numbers:
   REPORT(n_imm_reg);
   REPORT(n_imm_skp);
   REPORT(n_mat_new);
   REPORT(n_mat_old);
   REPORT(n_imm);
   REPORT(n_mat);
   
   return nll;
}

File: studies/size-based TMB/model_basis.R
library(gulf.data)
library(gulf.graphics)
library(TMB)
source("U:/TMB/TMB utilities.R")

setwd("C:/Users/SuretteTJ/Desktop/github/snow-crab-stock-assessment-2022/R/population model")
clc(); compile("model_basis.cpp")
dyn.load(dynlib("model_basis"))

# Prepare data:
load("C:/Users/SuretteTJ/Desktop/github/snow-crab-stock-assessment-2022/results/tables/size-frequencies males scs 2001-2022.rdata")
f <- f[, as.character(1:150), ]
f[,as.character(1:10),"immature"] <- 0
f[,as.character(121:150),"immature"] <- 0
f[,as.character(1:33),"skip"] <- 0
f[,as.character(121:150),"skip"] <- 0
f[,as.character(1:40),"recruit"] <- 0
f[,as.character(1:40),"residual"] <- 0
data <- list(x = as.numeric(dimnames(f)[[2]]),
             f_imm = f[,,"immature"] + f[,,"skip"],
             f_mat = f[,,"recruit"] + f[,,"residual"])

# Define instar basis functions:
mu_instars <- c(14.5, 20.5, 28.1, 37.8, 51.1, 68.0, 88.0, 106, 127)
names(mu_instars) <- 5:13
sigma_instars <- 0.08 * mu_instars
basis <- matrix(NA, nrow = length(mu_instars), ncol = length(data$x))
for (i in 1:length(mu_instars)){
   basis[i, ] <- sqrt(2*pi) * sigma_instars[i] * (pnorm(data$x + 0.5, mu_instars[i], sigma_instars[i]) - pnorm(data$x - 0.5, mu_instars[i], sigma_instars[i]))  
}
plot(c(0, 140), c(0, 1.1), type = "n", xlab = "", ylab = "", xaxs = "i", yaxs = "i")
grid()
for (i in 1:nrow(basis)){
   lines(data$x, basis[i,], lty = "dashed", col = rainbow(length(mu_instars))[i], lwd = 2)
   text(mu_instars[i], 1.035, as.roman(as.numeric(names(mu_instars[i]))))
}
#lines(data$x, apply(basis, 2, sum), lwd = 2, col = "grey30")
mtext("Carapace width (mm)", 1, 2.5, cex = 1.25)      
mtext("Value", 2, 2.5, cex = 1.25)  
years <- as.numeric(rownames(f))
box(col = "grey60")
data$basis <- basis

logit <- function(x) return(log(x/(1-x)))

# Define initial parameters: 
parameters <- list(intercept_growth = 0.276,             # Growth intercept parameter.
                   xp_growth = 38.2,                     # Growth pivot point parameter.
                   log_window_growth = 1.6,              # Growth transition window parameter.
                   slope_growth = c(0.32, 0.100),        # Growth slope parameters.
                   log_sigma_growth = log(0.135),        # Growth error parameter.
                   log_shape_selectivity = log(5),       # Shape parameter for cumulative gamma distribution function.
                   log_scale_selectivity = log(10),      # Scale parameter for cumulative gamma distribution function.
                   xp_maturity = c(55, 105),             # Maturation pivot point parameters.
                   log_window_maturity = c(-1.5, -0.1),  # Maturation transition window parameters.
                   logit_p_mix_maturity = 0.1,           # Mixing proportion between early and late maturation (logit-scale).
                   xp_skp = 70,                          # Skip-moulting pivot point parameter.
                   log_window_skp = -1.6,                # Skip-moulting transition window parameter.
                   logit_p_max_skp = 0,                  # Maximum skip-moulting probability (logit-scale).
                   xp_F = 98,                                       # Exploitation rate switching point.
                   log_window_F = -1,                               # Exploitation switching point window.
                   log_mu_F = logit(0.6),                           # Global exploitation rate.      
                   log_dev_F = 0.0 * rnorm(dim(data$f_imm)[1]),     # Annual deviations exploitation rate.
                   log_mu_M_imm = logit(0.25),                      # Global natural mortality rate.
                   log_dev_M_imm = rep(0, dim(data$f_imm)[1]),      # Annual deviations natural mortality rate.
                   log_mu_M_mat = logit(0.40),                      # Global natural mortality rate.
                   log_dev_M_mat = rep(0, dim(data$f_imm)[1]),      # Annual deviations natural mortality rate.
                   size_recruitment_mu = 14,                        # Mean recruitment size.
                   log_sigma_size_recruitment = -2,                 # Error for recruitment annual deviations.
                   size_recruitment = rep(0,length(years)),         # Annual deviations from recruitment instar mean size.
                   log_sigma_recruitment_instar = log(0.8),         # Error for recruitment instar.
                   log_scale_recruitment_mu = log(30000),           # Recruitment scale mean.
                   log_sigma_scale_recruitment = 10,                # Recruitment scale error.
                   log_scale_recruitment = rep(0,length(years)),    # Recruitment scale deviations.
                   log_N_imm_reg_beta = rep(1, nrow(data$basis)),   # Basis coefficients for immature crab for year 0.    
                   log_N_imm_skp_beta = rep(1, nrow(data$basis)-4), # Basis coefficients for immature skip-moulting crab for year 0.
                   log_N_mat_new_beta = rep(1, nrow(data$basis)-4), # Basis coefficients for new mature crab for year 0.
                   log_N_mat_old_beta = rep(1, nrow(data$basis)-4)) # Basis coefficients for old mature crab for year 0. 

# Set parameters:
parameters$log_sigma_dev_F = 0                            # Error for annual exploitation rate random effect.
parameters$log_sigma_dev_M_imm = 0                        # Error for annual mortality rate random effect.
parameters$log_sigma_dev_M_mat = 0                        # Error for annual mortality rate random effect.

# Error parameters:
parameters$log_r_imm <- 0.1
parameters$log_r_mat <- 0.1

# Adjust to better initial parameters values:
parameters$log_mu_F <- 0

parameters$log_scale_recruitment <- c(-0.8,-0.7,-0.2,0,0.2,-0.2,-0.3,0.2,0.1,-0.2,-0.2,-0.5,0.5,-0.2,0.3,0.2,-1.1,1.5,0.9,0.4,-1.2,-1.2)
parameters$log_scale_recruitment_mu <- 10.4
parameters$log_sigma_scale_recruitment <- -0.17

tolist <- function(x){
   str <- unique(names(x))
   r <- list()
   for (i in 1:length(str)){
      r[[i]] <- as.numeric(x[str[i] == names(x)])
   }
   names(r) <- str
   return(r)
}
update <- function(x, theta){
   theta <- tolist(theta)
   for (i in 1:length(theta)){
      x[[names(theta[i])]] <- theta[[i]]
   }
   return(x)
}

# parameters$size_recruitment <- 0.5*rnorm(length(years))

obj <- MakeADFun(data = data, parameters = parameters, DLL = "model_basis")
r <- obj$report()
image(r$p_recruitment)

map <- lapply(parameters, function(x) as.factor(NA * x))

# Fit dispersion parameters:
map$log_r_imm <- as.factor(1)
map$log_r_mat <- as.factor(1)
obj <- MakeADFun(data = data, parameters = parameters, map = map, DLL = "model_basis")
theta <- optim(obj$par, obj$fn, control = list(trace = 3, maxit = 1000))$par
parameters <- update(parameters, theta)

# Add initial immature:
map$log_N_imm_reg_beta <- as.factor(1:length(map$log_N_imm_reg_beta))
obj <- MakeADFun(data = data, parameters = parameters, map = map, DLL = "model_basis")
theta <- optim(obj$par, obj$fn, control = list(trace = 3, maxit = 2500))$par
parameters <- update(parameters, theta)

# Add initial skip:
map$log_N_imm_skp_beta <- as.factor(1:length(map$log_N_imm_skp_beta))
obj <- MakeADFun(data = data, parameters = parameters, map = map, DLL = "model_basis")
theta <- optim(obj$par, obj$fn, control = list(trace = 3, maxit = 2500))$par
parameters <- update(parameters, theta)

# Add initial new mature:
map$log_N_mat_new_beta <- as.factor(1:length(map$log_N_mat_new_beta))
obj <- MakeADFun(data = data, parameters = parameters, map = map, DLL = "model_basis")
theta <- optim(obj$par, obj$fn, control = list(trace = 3, maxit = 2500))$par
parameters <- update(parameters, theta)

# Add initial old mature:
map$log_N_mat_old_beta <- as.factor(1:length(map$log_N_mat_old_beta))
obj <- MakeADFun(data = data, parameters = parameters, map = map, DLL = "model_basis")
theta <- optim(obj$par, obj$fn, control = list(trace = 3, maxit = 3000))$par
parameters <- update(parameters, theta)

# Add recruitment:
map$size_recruitment_mu <- as.factor(1)                                     # Mean recruitment size.
map$log_scale_recruitment <- as.factor(1:length(map$log_scale_recruitment)) # Recruitment scale deviations.
map$log_scale_recruitment_mu <- as.factor(1)                                # Recruitment scale mean.
map$log_sigma_scale_recruitment <- as.factor(1)                             # Recruitment scale error.
#map$log_sigma_recruitment <- as.factor(1)
#size_recruitment_mu = 14,                        
#log_sigma_size_recruitment = -2,                 # Error for recruitment annual deviations.
#size_recruitment = rep(0,length(years)),         # Annual deviations from recruitment instar mean size.
#log_sigma_recruitment_instar = log(0.8),         # Error for recruitment instar.
#log_scale_recruitment_mu = log(30000),           
obj <- MakeADFun(data = data, parameters = parameters, map = map, DLL = "model_basis")
theta <- optim(obj$par, obj$fn, control = list(trace = 3, maxit = 2500))$par
parameters <- update(parameters, theta)

# Add maturity parameters:
#map$log_window_maturity <- as.factor(1:length(map$log_window_maturity))
map$logit_p_mix_maturity <- as.factor(1)
map$xp_maturity <- as.factor(1:length(map$xp_maturity))
obj <- MakeADFun(data = data, parameters = parameters, map = map, DLL = "model_basis")
theta <- optim(obj$par, obj$fn, control = list(trace = 3, maxit = 5000))$par
parameters <- update(parameters, theta)

# Add selectivity parameters:
map$log_shape_selectivity <- as.factor(1)
map$log_scale_selectivity <- as.factor(1)
obj <- MakeADFun(data = data, parameters = parameters, map = map, DLL = "model_basis")
theta <- optim(obj$par, obj$fn, control = list(trace = 3, maxit = 3000))$par
parameters <- update(parameters, theta)

# Add skip-moulters parameters:
map$xp_skp <- as.factor(1)
map$logit_p_max_skp <- as.factor(1)
obj <- MakeADFun(data = data, parameters = parameters, map = map, DLL = "model_basis")
theta <- optim(obj$par, obj$fn, control = list(trace = 3, maxit = 5000))$par
parameters <- update(parameters, theta)

# Add fishing mortality:
map$log_mu_F <- as.factor(1)
map$log_dev_F <- as.factor(1:length(map$log_dev_F))
map$log_sigma_dev_F <- as.factor(1)

map$log_window_F <- as.factor(1)
map$xp_F <- as.factor(1)
for (i in 1:5){
   obj <- MakeADFun(data = data, parameters = parameters, map = map, DLL = "model_basis")
   theta <- optim(obj$par, obj$fn, control = list(trace = 3, maxit = 5000))$par
   parameters <- update(parameters, theta)
}

# Add mortality parameters:
map$log_mu_M_imm <- as.factor(1)
map$log_dev_M_imm <- as.factor(1:length(map$log_dev_M_imm))
map$log_mu_M_mat <- as.factor(1)
map$log_dev_M_mat <- as.factor(1:length(map$log_dev_M_mat))
map$log_sigma_dev_M_imm <- as.factor(1)
map$log_sigma_dev_M_mat <- as.factor(1)
obj <- MakeADFun(data = data, parameters = parameters, map = map, DLL = "model_basis")
theta <- optim(obj$par, obj$fn, control = list(trace = 3, maxit = 5000))$par
parameters <- update(parameters, theta)

# Add growth parameters:
map$intercept_growth <- as.factor(1)
map$xp_growth
map$log_window_growth
map$slope_growth <- as.factor(1:length(map$slope_growth))
map$log_sigma_growth <- as.factor(1)
obj <- MakeADFun(data = data, parameters = parameters, map = map, DLL = "model_basis")
theta <- optim(obj$par, obj$fn, control = list(trace = 3, maxit = 5000))$par
parameters <- update(parameters, theta)

# Add recruitment sizes:
#map$log_sigma_recruitment <- as.factor(1)
#size_recruitment_mu = 14,                        
map$log_sigma_size_recruitment <- as.factor(1)                    # Error for recruitment annual deviations.
map$size_recruitment <- as.factor(1:length(map$size_recruitment)) # Annual deviations from recruitment instar mean size.
obj <- MakeADFun(data = data, parameters = parameters, map = map, DLL = "model_basis")
theta <- optim(obj$par, obj$fn, control = list(trace = 3, maxit = 2500))$par
parameters <- update(parameters, theta)

r <- obj$report()
dimnames(r$n_imm) <- list(year = (min(years)-1):max(years), cw = colnames(f))
dimnames(r$n_mat) <- list(year = (min(years)-1):max(years), cw = colnames(f))

random <- summary(obj)

gbarplot(exp(parameters$log_scale_recruitment_mu + parameters$log_scale_recruitment))
gbarplot(exp(parameters$log_N_imm_reg_beta))
gbarplot(exp(parameters$log_N_imm_skp_beta))
gbarplot(exp(parameters$log_N_mat_new_beta))
gbarplot(exp(parameters$log_N_mat_old_beta))
gbarplot(r$N_imm_reg[1,])
gbarplot(r$N_imm_skp[1,])
gbarplot(r$N_mat_new[1,])
gbarplot(r$N_mat_old[1,])

gbarplot(r$N_imm_reg[1,] + r$N_imm_skp[1,], data$x)
lines(data$x, r$n_imm[1,])
gbarplot(r$N_mat_new[1,] + r$N_mat_old[1,], data$x)
lines(data$x, r$n_mat[1,])

plot(r$p_selectivity, type = "l", lwd = 2)
plot(r$p_maturity, type = "l", lwd = 2)
plot(r$p_skp, type = "l", lwd = 2)
gbarplot(1/(1+ exp(-exp(-parameters$log_mu_F - parameters$log_dev_F))))
plot(r$mu_growth, type = "l", lwd = 2)
gbarplot(1/(1+ exp(-parameters$log_mu_M_imm - parameters$log_dev_M_imm)))
gbarplot(1/(1+ exp(-parameters$log_mu_M_mat - parameters$log_dev_M_mat)))

#F <- exp(parameters$log_mu_F + parameters$log_dev_F[i]) / (1+exp(-exp(parameters$log_window_F) * (data$x - parameters$xp_F)))
gbarplot(r$F[1,], data$x)

clg()
#windows()
r <- obj$report()
dimnames(r$n_imm) <- list(year = (min(years)-1):max(years), cw = colnames(f))
dimnames(r$n_mat) <- list(year = (min(years)-1):max(years), cw = colnames(f))

m <- kronecker(matrix(1:21, ncol = 3), matrix(1, ncol = 5, nrow = 5))
m <- rbind(0, cbind(0, 0, m, 0), 0, 0)
layout(m)
par(mar = c(0,0,0,0))

for (i in 2:length(years)){
   year <- as.character(years[i])
   gbarplot(cbind(r$n_imm[year, ], r$n_mat[year, ]) , as.numeric(colnames(f)), 
            legend = FALSE, grid = TRUE, xaxs = "i", xlim = c(0, 140), yaxs = "i", # ylim = c(0, 800),  
            xaxt = "n", yaxt = "n")
   lines(data$x, data$f_imm[year, ] + data$f_mat[year, ], lwd = 2, col = "blue")
   lines(data$x, data$f_imm[year, ], lwd = 2, col = "red")
   
   if (i %in% 2:8) axis(2)
   if (i %in% (c(7, 14, 21)+1)) axis(1)
   text(par("usr")[1] + 0.8 * diff(par("usr")[1:2]),
        par("usr")[3] + 0.8 * diff(par("usr")[3:4]), year)
   
   vline(c(9.4, 14, 20.3, 28.5, 38), lty = "dashed")
   box(col = "grey60")
}

random <- c("log_dev_F", "log_dev_M_imm", "log_dev_M_mat", "log_scale_recruitment", "size_recruitment")

obj <- MakeADFun(data = data, parameters = parameters, random = random, map = map, DLL = "model_basis")
theta <- optim(obj$par, obj$fn, control = list(trace = 3, maxit = 1000))$par
parameters <- update(parameters, theta)

rep <- sdreport(obj)
theta <- c(theta, summary(rep, "random")[,1])
parameters <- update(parameters, theta)
# parameters <- update.parameters(parameters, summary(rep, "fixed"), summary(rep, "random"))



